### PR TITLE
[Bugfix][Spec] Fix multi-layer EAGLE DRAFT_EXTEND_V2 attn-TP logprob metadata capture

### LINKED ISSUE #26416: [Bug] multi-layer EAGLE DRAFT_EXTEND_V2 captures bs instead of num_tokens for attn-TP logprob metadata
### Summary

`MultiLayerEagleDraftExtendCudaGraphRunner.get_forward_batch()` captures wrong DP/logprob metadata in the `require_attn_tp_gather` branch for `ForwardMode.DRAFT_EXTEND_V2`.

For the branch:

- `require_mlp_tp_gather == False`
- `require_attn_tp_gather == True`
- `forward_mode == DRAFT_EXTEND_V2`
- `speculative_num_steps >= 2`

current `main` initializes:

```python
global_num_tokens_for_logprob_gpu = [bs]
```

but DRAFT_EXTEND_V2 computes logits for all draft-extend tokens, so this should be `num_tokens = bs * self.num_tokens_per_bs`. The replay path already fills `bs * self.num_tokens_per_bs`, and the single-layer `eagle_draft_extend_cuda_graph_runner.py` uses `num_tokens_for_logprob = num_tokens` for DRAFT_EXTEND_V2.

Related open PR found before filing: #26016 appears to touch the broader `require_attn_tp_gather=True` path, but I did not find an open issue for this exact captured metadata mismatch.

### Environment

- SGLang source commit: `47617cc4df1e5c249d11ea5970ce26c32c8757af` (`main`)
- GPU: NVIDIA A100-SXM4-40GB
- Driver: 580.159.03
- CUDA runtime reported by PyTorch: 12.9
- PyTorch: 2.9.1+cu129
- GCP VM: `a2-highgpu-1g` Spot in `us-west4-b`
- Image: `pytorch-2-9-cu129-ubuntu-2204-nvidia-580-v20260526`

### Focused repro

I used a focused CUDA/PyTorch instrumentation script that imports and executes the real `MultiLayerEagleDraftExtendCudaGraphRunner.get_forward_batch()` implementation from the source tree, with unrelated SGLang modules stubbed out. This avoids needing a large multi-layer EAGLE model while still exercising the actual branch and CUDA tensor mutation.

Exact command on the A100 VM:

```bash
cd ~/sglang-repro
python3 scripts/repro_multi_layer_eagle_attn_tp_gather_metadata.py --bs 2 --speculative-num-steps 2
```

The script sets `step=1`, so `num_tokens_per_bs = speculative_num_steps + 1 + step = 4`, and `expected_num_tokens = bs * num_tokens_per_bs = 8`.

### Current main result

```json
{
  "branch_predicates": {
    "forward_mode": "DRAFT_EXTEND_V2",
    "require_attn_tp_gather": true,
    "require_mlp_tp_gather": false,
    "speculative_num_steps": 2
  },
  "bs": 2,
  "capture_metadata": {
    "forward_batch_global_dp_buffer_len": 8,
    "forward_batch_logits_rows": 8,
    "global_num_tokens_for_logprob_gpu": 2,
    "global_num_tokens_gpu": 8
  },
  "cuda_runtime": "12.9",
  "device": "NVIDIA A100-SXM4-40GB",
  "expected_num_tokens": 8,
  "num_tokens_per_bs": 4,
  "replay_metadata_fill": {
    "cuda_graph_readback_after_replay_fill": 8,
    "global_num_tokens_for_logprob_gpu": 8,
    "global_num_tokens_gpu": 8
  },
  "torch_version": "2.9.1+cu129",
  "wrong_capture_logprob_metadata": true
}
```

This shows capture initializes `global_num_tokens_for_logprob_gpu=2` (`bs`) while the expected DRAFT_EXTEND_V2 token count and replay metadata are both `8`.

### Patched result

After applying the one-line change below and rerunning the same command:

```json
{
  "capture_metadata": {
    "forward_batch_global_dp_buffer_len": 8,
    "forward_batch_logits_rows": 8,
    "global_num_tokens_for_logprob_gpu": 8,
    "global_num_tokens_gpu": 8
  },
  "expected_num_tokens": 8,
  "replay_metadata_fill": {
    "cuda_graph_readback_after_replay_fill": 8,
    "global_num_tokens_for_logprob_gpu": 8,
    "global_num_tokens_gpu": 8
  },
  "wrong_capture_logprob_metadata": false
}
```

### Minimal diff

```diff
diff --git a/python/sglang/srt/speculative/multi_layer_eagle_draft_extend_cuda_graph_runner.py b/python/sglang/srt/speculative/multi_layer_eagle_draft_extend_cuda_graph_runner.py
@@
             buffers.global_num_tokens_for_logprob_gpu.copy_(
                 torch.tensor(
-                    [bs],
+                    [num_tokens],
                     dtype=torch.int32,
                     device=buffers.input_ids.device,
                 )
```

### Expected behavior

For DRAFT_EXTEND_V2 in the attention-TP gather branch, capture should initialize `global_num_tokens_for_logprob_gpu` with the number of logits/logprob hidden states (`num_tokens`), matching:

- `next_token_logits_buffer[:num_tokens]`
- `global_dp_buffer_len = num_tokens`
- replay's `buffers.global_num_tokens_for_logprob_gpu.fill_(bs * self.num_tokens_per_bs)`
- single-layer EAGLE draft-extend CUDA graph runner behavior

### PR DESCRIPTION
**Pull request by Sunt-ing**
## Motivation

Fixes #26416 (reported by @glaziermag).

`MultiLayerEagleDraftExtendCudaGraphRunner.get_forward_batch()` set `global_num_tokens_for_logprob_cpu = [bs]` in the `require_attn_tp_gather` branch. This runner always runs `ForwardMode.DRAFT_EXTEND_V2`, which produces logits for all draft-extend tokens, so the logprob metadata should be `num_tokens = bs * num_tokens_per_bs`, not `bs`.

The `[bs]` value disagreed with everything else on the same path: `global_num_tokens_cpu` (`[num_tokens]`), `global_dp_buffer_len` and `next_token_logits_buffer[:num_tokens]`, the replay fill (`global_num_tokens_for_logprob_gpu.fill_(bs * num_tokens_per_bs)`), the single-layer `eagle_draft_extend_cuda_graph_runner` (uses `num_tokens` for DRAFT_EXTEND_V2), and the sibling `require_mlp_tp_gather` branch in this same runner (already `[num_tokens]`).

## Modifications

`python/sglang/srt/speculative/multi_layer_eagle_draft_extend_cuda_graph_runner.py`: in the `require_attn_tp_gather` branch of `get_forward_batch()`, set `global_num_tokens_for_logprob_cpu = [num_tokens]` (was `[bs]`), aligning it with the `require_mlp_tp_gather` branch. No other branch changes.

## Testing

One-line parity correction; the existing EAGLE draft-extend coverage under `test/registered/spec/eagle/` exercises this runner. The captured `global_num_tokens_for_logprob` goes from `2` (wrong, `bs`) to `8`, matching `global_num_tokens`, `global_dp_buffer_len`, and the replay fill (`bs * num_tokens_per_bs`); the `require_mlp_tp_gather` branch was already `8`.

<details>
<summary>repro (env + script + before/after)</summary>

Environment:
- GPU: NVIDIA A800-80GB (sm80)
- SGLang: `main`
- A full multi-GPU attention-TP EAGLE run is not available on a single GPU, so this drives the real `get_forward_batch()` directly. The runner's `__init__` needs a full eagle worker, so the instance is built with `object.__new__` and only the attributes `get_forward_batch()` reads are set, with real GPU input buffers.

Script (`bs=2`, `speculative_num_steps=2`, `step=1` so `num_tokens_per_bs=4`, `num_tokens=8`):

```python
import torch
from sglang.srt.model_executor.forward_batch_info import ForwardMode
from sglang.srt.speculative.multi_layer_eagle_draft_extend_cuda_graph_runner import (
    MultiLayerEagleDraftExtendCudaGraphRunner as Runner,
    MultiLayerEagleDraftExtendInputBuffers as Buffers,
)

dev = "cuda"
bs, steps, step = 2, 2, 1
num_tokens_per_bs = steps + 1 + step          # 4
num_tokens = bs * num_tokens_per_bs           # 8


class _Algo:
    def is_standalone(self):
        return False


class _ModelRunner:
    spec_algorithm = _Algo()


mt = num_tokens
buffers = Buffers(
    input_ids=torch.zeros(mt, dtype=torch.int64, device=dev),
    out_cache_loc=torch.zeros(mt, dtype=torch.int64, device=dev),
    positions=torch.zeros(mt, dtype=torch.int64, device=dev),
    seq_lens=torch.full((bs,), 8, dtype=torch.int32, device=dev),
    seq_lens_cpu=torch.full((bs,), 8, dtype=torch.int32),
    req_pool_indices=torch.arange(bs, dtype=torch.int32, device=dev),
    num_correct_drafts=torch.zeros(bs, dtype=torch.int32, device=dev),
    num_accept_tokens=torch.zeros(bs, dtype=torch.int32, device=dev),
    extend_seq_lens=torch.full((bs,), num_tokens_per_bs, dtype=torch.int32, device=dev),
    extend_start_loc=torch.arange(0, mt, num_tokens_per_bs, dtype=torch.int32, device=dev),
    mrope_positions=torch.zeros((3, mt), dtype=torch.int64, device=dev),
    hidden_states=torch.zeros((mt, 16), dtype=torch.bfloat16, device=dev),
    next_token_logits_buffer=torch.zeros((mt, 32), dtype=torch.float, device=dev),
    global_num_tokens_gpu=torch.zeros((1,), dtype=torch.int32, device=dev),
    global_num_tokens_for_logprob_gpu=torch.zeros((1,), dtype=torch.int32, device=dev),
)

r = object.__new__(Runner)
r.buffers = buffers
r.num_tokens_per_bs = num_tokens_per_bs
r.forward_mode = ForwardMode.DRAFT_EXTEND_V2
r.require_mlp_tp_gather = False
r.require_attn_tp_gather = True
r.dp_size = 1
r.padded_static_len = -1
r.extend_seq_lens_cpu = [num_tokens_per_bs] * bs
r.model_runner = _ModelRunner()

r.get_forward_batch(bs)
print("expected num_tokens          :", num_tokens)
print("global_num_tokens            :", int(r.buffers.global_num_tokens_gpu[0]))
print("global_num_tokens_for_logprob:", int(r.buffers.global_num_tokens_for_logprob_gpu[0]))
```

Before the fix:
```
expected num_tokens          : 8
global_num_tokens            : 8
global_num_tokens_for_logprob: 2
```

After the fix:
```
expected num_tokens          : 8
global_num_tokens            : 8
global_num_tokens_for_logprob: 8
```

</details>

## Checklist

- [x] Format the code with pre-commit.

cc @hnyls2002 @ch-wan

---
### CI States

Latest PR Test (Base): :x: [Run #27402134362](https://github.com/sgl-project/sglang/actions/runs/27402134362)
Latest PR Test (Extra): :x: [Run #27402134279](https://github.com/sgl-project/sglang/actions/runs/27402134279)

### CODE REVIEW

**Comment by glaziermag** [comment]
Thank you @Sunt-ing for fixing this and for the clear repro and validation notes. I appreciate it.

**Comment by hnyls2002** [comment]
/rerun-test test_mimo_v2.py test_mimo_v2_flash.py test_step3p5_flash_chain_mtp.py

**Review by hnyls2002 — APPROVED** [review]
